src/model.c

```diff
@@ -255,6 +287,115 @@ static bool parse_int(char **buffer, int *i)
     return true;
 }
 
+struct model *model_load_from_ply(const char *fname)
+{
+    FILE *fp = fopen(fname, "r");
+    if (!fp)
+    {
+        fprintf(stderr, "ERROR: failed to load file \"%s\".\n", fname);
+        return NULL;
+    }
+
+    // Create a new model
+    struct model *model = model_init();
+
+    // Read each line of the file
+    char buffer[256];
+
+    int vertex_count = 0;
+    int face_count = 0;
+
+    while (fgets(buffer, sizeof(buffer), fp))
+    {
+        char *p = buffer;
+        while (*p)
+        {
+            if (*p == '\n' || *p == '\r')
+                *p = '\0';
+            if (*p == '\t')
+                *p = ' ';
+            p++;
+        }
+
+        char *bufferp = buffer;
+        char *instr = str_chop_skip_empty(&bufferp, " ");
+
+        if (!instr || instr[0] == '#')
+            continue;
+
+        if (strcmp(instr, "element") == 0)
+        {
+            char *type = str_chop_skip_empty(&bufferp, " ");
+            int count = 0;
+            parse_int(&bufferp, &count);
+
+            if (strcmp(type, "vertex") == 0)
+            {
+                model_reserve_vertexes(model, count);
+                vertex_count = count;
+            }
+            else if (strcmp(type, "face") == 0)
+            {
+                model_reserve_faces(model, count);
+                face_count = count;
+            }
+        }
+        else if (strcmp(instr, "property") == 0)
+        {
+            // Skip property declarations
+        }
+        else if (strcmp(instr, "end_header") == 0)
+        {
+            // End of header, start reading data
+            break;
+        }
+    }
+
+    for (int i = 0; i < vertex_count; i++)
+    {
+        float f1, f2, f3;
+
+        if (!fgets(buffer, sizeof(buffer), fp) ||
+            sscanf(buffer, "%f %f %f", &f1, &f2, &f3) != 3)
+        {
+            fprintf(stderr, "ERROR: invalid vertex data.\n");
+            fclose(fp);
+            model_free(model);
+            return NULL;
+        }
+
+        vec3 vec;
+        vec.x = f1;
+        vec.y = f2;
+        vec.z = f3;
+
+        model_add_vertex(model, vec);
+    }
+
+    for (int i = 0; i < face_count; i++)
+    {
+        int count;
+        int idxs[3];
+
+        if (!fgets(buffer, sizeof(buffer), fp) ||
+            sscanf(buffer, "%d %d %d %d", &count, &idxs[0], &idxs[1], &idxs[2]) != 4 ||
+            count != 3)
+        {
+            fprintf(stderr, "ERROR: invalid face data.\n");
+            fclose(fp);
+            model_free(model);
+            return NULL;
+        }
```

Comment:
My understanding is that with this implementation we would not be supporting faces with more than 3 vertexes. I think we should, since triangularization is fairly easy to do, under the common assumption that the vertices are within the same plane and the face is convex. 

We already do that for OBJ. Notice how if a face has `n` vertices, `n - 2` triangles are added to the model with model_add_face().

Reply:
I changed this to allow for more than 3 vertices on each face.